report/5.Solution/Scheduling.tex

```diff
@@ -84,14 +84,83 @@ \section{Scheduling}\label{solution:scheduling}
 
-\todo{Update handle lasers and stop times when they are implemented}
+\subsection{Fixed Priority Scheduling}
+However, as the project progressed, parts of the machine intelligence execution was moved to the NXT for execution.
+This lead to tasks that took longer than 1 millisecond to execute, rendering the use of cyclic executive infeasible, as described in Subsection~\cite{Design:Scheduling}. 
+Due to this, the fixed-priority scheduling model was instead utilized. 
+
+After having added machine learning to the \texttt{Move\_Motors} task the execution times are as follows:
+
+\begin{table}[tbp]
+\begin{tabular}{lll}
+\textbf{Task name}  & \textbf{Execution time} 	& \textbf{Priority}\\
+Keep\_USB\_Alive    & 14                        & 6    \\
+Receive\_Data       & 13                        & 4   \\
+Move\_Motors        & --                        & 2   \\
+Handle\_Lasers      & 30                        & 3   \\
+Update\_Display     & 480                       & 1   \\                     
+Background\_Task    & -                        	& 5  \\                     
+\end{tabular}
+\end{table}\label{table:executionTimes}
+\todo{Add execution time of move motors and bg task}
+
+As the cyclic executive approach allowed running the calibration before entering the first major cycle, this was easily implemented.
+However, this does not work with the fixed-priority scheduling model, as the tasks are repeatedly run according to their tasks.
+To fix this, the \texttt{Background\_Task} was introduced, which has a high enough priority to preempt the other tasks on the system to allow calibration to run, while still keeping the USB connection alive.
+
+\begin{lstlisting}[language=CSharp,label={lst:backgroundTask},caption={Background task}]
+TASK(background_task) {
+    if(!calibrated){
+        show_init_screen();
+        while(!get_status_code(&current_status, 0));
+
+        if (current_status == READY_FOR_CALIBRATION) {
+            calibrate(false);
+        }
+        calibrated = true;
+    }
+    if (current_status == DISCONNECTED_REQ) {
+        stop();
+    }
+    TerminateTask();
+}
+\end{lstlisting}
+
+As seen on Snippet~\ref{lst:backgroundTask}, the task checks if the boolean value calibrated has been changed to true, meaning that the calibration has already be done.
+If this is not the case, the background task ensures that the status of the motors and USB are set as disconnected using the \texttt{stop} function.
+
+In the cyclic executive model, it was ensured that the \texttt{keep\_USB\_alive} task was executed once every millisecond by running it at the beginning at each cycle with a cycle length of 1 millisecond.
+With the FPS model, this is instead handled by assigning the task the highest possible priority, ensuring that it will run regardless of whatever else is currently being executed using preemption if needed.
```

Comment:
```suggestion
With the FPS model, this is instead done by assigning the \texttt{keep\_USB\_alive} task with the highest possible priority.
This ensures that it will run ahead of all other tasks, using preemption if needed.
```